goalie/go_mesh_seq.py

```diff
@@ -65,6 +68,52 @@ def forms(self):
             )
         return self._forms
 
+    @PETSc.Log.EventDecorator()
+    def _detect_changing_coefficients(self, export_idx):
+        """
+        Detect whether coefficients other than the solution in the variational forms
+        change over time. If they do, store the changed coefficients so we can update
+        them in :meth:`~.GoalOrientedMeshSeq.indicate_errors`.
+
+        Changed coefficients are stored in a dictionary with the following structure:
+        ``{field: {coeff_idx: {export_timestep_idx: coefficient}}}``, where
+        ``coefficient=forms[field].coefficients()[coeff_idx]`` at export timestep
+        ``export_timestep_idx``.
+
+        :arg export_idx: index of the current export timestep within the subinterval
+        :type export_idx: :class:`int`
+        """
+        # Copy coefficients at subinterval's first export timestep
+        if export_idx == 0:
+            self._prev_form_coeffs = {
+                field: deepcopy(form.coefficients())
+                for field, form in self.forms.items()
+            }
+            self._changed_form_coeffs = {field: {} for field in self.fields}
+        # Store coefficients that have changed since the previous export timestep
+        else:
+            for field in self.fields:
+                # Coefficients at the current timestep
+                coeffs = self.forms[field].coefficients()
+                for coeff_idx, (coeff, init_coeff) in enumerate(
+                    zip(coeffs, self._prev_form_coeffs[field])
+                ):
+                    # Skip solution fields since they are stored separately
+                    if coeff.name().split("_old")[0] in self.time_partition.field_names:
+                        continue
+                    if not np.array_equal(
+                        coeff.vector().array(), init_coeff.vector().array()
+                    ):
```

Comment:
How come you use `array_equal` here rather than `allclose`? The coefficient values will generally be real values rather than integers, no?

Reply:
Well if a coefficient isn't changing in time it should have exactly the same data, no? But I changed it to `allclose` in https://github.com/mesh-adaptation/goalie/pull/259/commits/a39154a501141d09adbccdedd621e124597b8ea9 since I guess I'm missing something :)

Reply:
True. But I think using `allclose` when handling floating point numbers should make things more robust.

Reply:
Makes sense :) thanks!